cirq/protocols/apply_mixture_protocol_test.py

```diff
@@ -0,0 +1,330 @@
+
+
+def test_apply_mixture_no_protocols_implemented_default():
+
+    class NoProtocols:
+        pass
+
+    args = cirq.ApplyMixtureArgs(target_tensor=np.eye(2),
+                                 left_axes=[0],
+                                 right_axes=[1],
+                                 out_buffer=None,
+                                 auxiliary_buffer0=None,
+                                 auxiliary_buffer1=None)
+    result = cirq.apply_mixture(NoProtocols(), args, default='cirq')
+    assert result == 'cirq'
+
+
+def test_apply_mixture_unitary():
+    m = np.diag([1, 1j])
+
+    shape = (2, 2, 2, 2)
+    rho = np.ones(shape, dtype=np.complex128)
+
+    class HasUnitary:
+
+        def _unitary_(self) -> np.ndarray:
+            return m
+
+    class HasUnitaryButReturnsNotImplemented(HasUnitary):
+
+        def _apply_unitary_(self, args: cirq.ApplyMixtureArgs):
+            return NotImplemented
+
+    for val in (HasUnitary(), HasUnitaryButReturnsNotImplemented()):
+        result = apply_mixture(val,
+                               rho,
+                               left_axes=[1],
+                               right_axes=[3],
+                               assert_result_is_out_buf=False)
+
+        np.testing.assert_almost_equal(
+            result,
+            np.reshape(np.outer([1, 1j, 1, 1j], [1, -1j, 1, -1j]), shape),
+        )
+
+
+def test_apply_mixture_apply_unitary():
+    shape = (2, 2, 2, 2)
+    rho = np.ones(shape, dtype=np.complex128)
+
+    class HasApplyUnitaryOutputInBuffer:
+
+        def _apply_unitary_(self, args: cirq.ApplyUnitaryArgs) -> np.ndarray:
+            zero = args.subspace_index(0)
+            one = args.subspace_index(1)
+            args.available_buffer[zero] = args.target_tensor[zero]
+            args.available_buffer[one] = 1j * args.target_tensor[one]
+            return args.available_buffer
+
+    class HasApplyUnitaryMutateInline:
+
+        def _apply_unitary_(self, args: cirq.ApplyUnitaryArgs) -> np.ndarray:
+            one = args.subspace_index(1)
+            args.target_tensor[one] *= 1j
+            return args.target_tensor
+
+    for val in (HasApplyUnitaryOutputInBuffer(), HasApplyUnitaryMutateInline()):
+        result = apply_mixture(val,
+                               rho,
+                               left_axes=[1],
+                               right_axes=[3],
+                               assert_result_is_out_buf=False)
+        np.testing.assert_almost_equal(
+            result, np.reshape(np.outer([1, 1j, 1, 1j], [1, -1j, 1, -1j]),
+                               shape))
+
+
+def test_apply_mixture_apply_unitary_not_implemented():
+
+    class ApplyUnitaryNotImplemeneted:
+
+        def _apply_unitary_(self, args: cirq.ApplyUnitaryArgs):
+            return NotImplemented
+
+    rho = np.ones((2, 2, 2, 2), dtype=np.complex128)
+    out_buf, aux_buf0, aux_buf1 = make_buffers((2, 2, 2, 2), dtype=rho.dtype)
+
+    with pytest.raises(TypeError):
```

Comment:
Include a `match="..."` argument for all `pytest.raises`, to avoid false negatives.

Reply:
added